### DISCUSSION TITLE
`as not` type-matching pattern

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#6873] [Language Ideas]
There are some rare cases where you want a variable to not be of some specified types. The `as not` pattern can help filtering out the types that are not desirable for a value.

The pattern can be used combining multiple types like `as not X and not Y and not Z`. The types must be combined with `and not`, any other combination is not permitted (`or`, `and` without `not`).

The resulting value of the `as not` operator is the same value as the given expression's on the left side of the `as not` expression. If the value is of any of the excluded types, `null` is returned instead. More formally, given the expression `expr` which has a type `T`, the expression `expr as not X and not Y` returns `T`.

# Example
```csharp
T Filter<T>(T value)
    where T : class
{
    return value as not string and not List<int>;
}
```
equivalent to
```csharp
T Filter<T>(T value)
    where T : class
{
    return value is string or List<int> ? null : value;
}
```

**Comment by HaloFour** [comment]
I think the bigger ask here is for `as` to work with patterns at all, which it currently does not.

**Comment by svick** [comment]
If, as you say, this is only useful in rare cases, why does it need a syntax that's slightly shorter than the existing alternative?
